Add Socket.IO v2 support

SeinopSys opened the pull request:
My site depends on this library for communicating with a separately running Node.js socket.io server, and I noticed that as of version 2.0.0 of socket.io Elephant.IO is failing to connect to the socket.io server and sending messages.

Therefore I dug into the code and messed around with it until the Node.js server was logging that it got the messages again. I can't guarantee with 100% certainty that this change set will permanently fix the issues as I don't have deep understanding of WebSockets or socket.io's API, but for the time being I managed to send messages with it and I consider that to be better than an error message saying 
> in_array expects parameter 2 to be array, null given in .../src/Engine/SocketIO/Version1X.php on line [172](https://github.com/Wisembly/elephant.io/blob/master/src/Engine/SocketIO/Version1X.php#L172)

I'd highly appreciate if this fix could be merged for the sake of anyone else trying to upgrade.

Taluu commented:
Basically, if I understand it correctly, it's just that the sent information has some more content after the closing `}` when upgrading the transport ? IMO it would be better to either just extend the Socket 1.X class, and replace the upgrade method, or even better, do what you did with the closing brace on Socket 1.x.

Unless I missed something else ?

Taluu commented on src/Engine/SocketIO/Version2X.php:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @author Baptiste Clavié <baptiste@wisembly.com>
+ * @link https://tools.ietf.org/html/rfc6455#section-5.2 Websocket's RFC
+ */
```
the comment is not right

SeinopSys commented:
I'm currently checking whether the commit I pushed works, I tried to move what I thought I could into 1X and leaving only what must stay in 2X.

Taluu commented on src/Engine/SocketIO/Version1X.php:
```diff
@@ -193,7 +195,7 @@ private function upgradeTransport()
 
         $url = sprintf('/%s/?%s', trim($this->url['path'], '/'), http_build_query($query));
-        $key = base64_encode(sha1(uniqid(mt_rand(), true), true));
+        $key = base64_encode(random_bytes(20));
```
that's not good, it is php7+ only (we don't have the polyfill too, and I don't want to introduce it yet), and this lib supports 5.4. Supporting 7 would be changing some stuff (not that I'm against it, but if we are gonna make it to 7.0, everything should be adapted)

Taluu replied:
(I meant the call to `random_bytes`)

SeinopSys commented:
@Taluu See the updated version. Apparently the break's main reason was the previously mentioned extra data after the JSON plus the `Sec-WebSocket-Key` being too long. Making it shorter was necessary.

Taluu commented on src/Engine/SocketIO/Version1X.php:
```diff
@@ -187,15 +187,17 @@ protected function handshake()
+
         $url = sprintf('/%s/?%s', trim($this->url['path'], '/'), http_build_query($query));
-        $key = base64_encode(random_bytes(20));
+        $key = base64_encode(openssl_random_pseudo_bytes(16));
```
Not good either, as it requires openssl. If you absolutely need 20 bytes encoded, you can pass the second argument of sha1 to `true`, it will generate you a binary digest with a length of 20 bytes,

Taluu commented on src/Engine/SocketIO/Version1X.php:
```diff
@@ -187,15 +187,17 @@ protected function handshake()
                   'transport' => static::TRANSPORT_WEBSOCKET];
 
+	    if ($this->options['version'] === 2)
+	        $query['use_b64'] = $this->options['use_b64'];
```
missing braces

Taluu replied:
(PSR and all that)

Taluu commented:
Hum, I see for the header length. You could always crop it to the maximum number of chars, if on version 2 then I guess ?

Taluu commented on src/Engine/SocketIO/Version1X.php:
```diff
@@ -167,7 +167,9 @@ protected function handshake()
 
-        $decoded = json_decode(substr($result, strpos($result, '{')), true);
+	    $open_curly_at = strpos($result, '{');
+	    $todecode = substr($result, $open_curly_at, strrpos($result, '}')-$open_curly_at+1);
```
tab to space please (code style)

SeinopSys commented:
Fixed the tab/space inconsistency & added a check to cut the hash length down for v2. `true` was already set for the function call, but I removed to so I could use `substr` instead.

Taluu commented:
Still missing a brace of two on "simple conditions" but I'll merge it anyway, Thanks @SeinopSys 

(I'll make a release too)

Taluu commented:
Release done : https://github.com/Wisembly/elephant.io/releases/tag/v3.3.0

SeinopSys commented:
Sorry about that, leaving them off has become a habit of mine. I'm glad I could help keep the lib flowing for a little longer.

SeinopSys commented:
I've just noticed that I left a few `use` statements in the 2X engine file, not sure if it's a huge issue or not but they probably should not be in there.

Taluu commented:
I'm pretty used to miss them, so some others it's not that big of a deal. Will fix if I ever have to fix something in Elephant.
